src/csync/csync_exclude.cpp

```diff
@@ -320,106 +322,249 @@ static CSYNC_EXCLUDE_TYPE _csync_excluded_common(c_strlist_t *excludes, const ch
-            }
-        }
-        /* Check if the pattern applies to pathes only. */
-        if (pattern[strlen(pattern)-1] == '/') {
-            if (!check_leading_dirs && filetype == CSYNC_FTW_TYPE_FILE) {
-                continue;
-            }
-            match_dirs_only = true;
-            pattern[strlen(pattern)-1] = '\0'; /* Cut off the slash */
-        }
+    return match;
+}
 
-        /* check if the pattern contains a / and if, compare to the whole path */
-        if (strchr(pattern, '/')) {
-            rc = csync_fnmatch(pattern, path, FNM_PATHNAME);
-            if( rc == 0 ) {
-                match = type;
-            }
-            /* if the pattern requires a dir, but path is not, its still not excluded. */
-            if (match_dirs_only && filetype != CSYNC_FTW_TYPE_DIR) {
-                match = CSYNC_NOT_EXCLUDED;
-            }
-        }
 
-        /* if still not excluded, check each component and leading directory of the path */
-        if (match == CSYNC_NOT_EXCLUDED && check_leading_dirs) {
-            size_t j = 0;
-            if (match_dirs_only && filetype == CSYNC_FTW_TYPE_FILE) {
-                j = 1; // skip the first entry, which is bname
-            }
-            for (; j < path_components->count; ++j) {
-                rc = csync_fnmatch(pattern, path_components->vector[j], 0);
-                if (rc == 0) {
-                    match = type;
-                    break;
-                }
-            }
-        } else if (match == CSYNC_NOT_EXCLUDED && !check_leading_dirs) {
-            rc = csync_fnmatch(pattern, bname, 0);
-            if (rc == 0) {
-                match = type;
-            }
+static QString convertToFullRegexpSyntax(QString exclude)
+{
+    QString s = QRegularExpression::escape(exclude).replace("\\*", "[^/]*").replace("?", "[^/]");
+    // Anchoring to the beginning if pattern starts with a /
+    if (exclude.startsWith("/")) {
+        s.prepend("^");
+    } else {
+        // we don't want "foo/bar" to match "afoo/bar"
+        s.prepend("/");
+    }
+    if (!exclude.endsWith("/"))
+        s.append("/?");
+    //qDebug() << "Converted pattern" << exclude << "to regex" << s;
+    return s;
+}
+
+// FIXME: rename to traversal_prepare
+void csync_exclude_load_finalize(CSYNC *ctx)
+{
+    // env QT_ENABLE_REGEXP_JIT 1
+
+    QString _exclude_traversel_regexp_exclude;
+    QString _exclude_traversel_regexp_exclude_and_remove;
+    for (unsigned int i = 0; i < ctx->excludes->count; i++) {
+        char *exclude = ctx->excludes->vector[i];
+        QString *builderToUse = & _exclude_traversel_regexp_exclude;
+        if (exclude[0] == '\n') continue; // empty line
+        if (exclude[0] == '\r') continue; // empty line
+
+        /* If an exclude entry contains a slash, we use the C-style codepath without QRegularEpression */
+        if (strchr(exclude, '/')) {
+            _csync_exclude_add(&ctx->parsed_traversal_excludes.list_patterns_with_slashes, exclude);
```

Comment:
I think special casing these at first to get the patch done quicker and with less room for mistakes is nice.

Reply:
@guruz Also, `list_patterns_with_slashes` needs to be wiped in this function to make it not grow longer and longer.